src/d3a/models/strategy/commercial_producer.py

```diff
@@ -60,15 +62,28 @@ def event_market_cycle(self):
 
     def offer_energy(self, market):
         energy_rate = self.energy_rate[market.time_slot]
-        offer = market.offer(
-            self.energy_per_slot_kWh * energy_rate,
-            self.energy_per_slot_kWh,
-            self.owner.name,
-            original_offer_price=self.energy_per_slot_kWh * energy_rate,
-            seller_origin=self.owner.name
-        )
-
-        self.offers.post(offer, market.id)
+        try:
+            offer = market.offer(
+                self.energy_per_slot_kWh * energy_rate,
+                self.energy_per_slot_kWh,
+                self.owner.name,
+                original_offer_price=self.energy_per_slot_kWh * energy_rate,
+                seller_origin=self.owner.name
+            )
+
+            self.offers.post(offer, market.id)
+        except MarketException:
+            logging.error(f"Offer posted with negative energy rate {energy_rate}."
+                          f"Posting offer with zero energy rate instead.")
+            offer = market.offer(
+                0.0,
+                self.energy_per_slot_kWh,
+                self.owner.name,
+                original_offer_price=0.0,
+                seller_origin=self.owner.name
+            )
```

Comment:
I would rather not self assume and post offer with zero price in case the external agent has placed an offer with negative rate, the reason is that it may be he/she wants to sell energy at 10 cents but due to any of his code bug end up placing offer at -10cents, its better to only throw an exception back to researcher so he can adapt his script accordingly rather him wonder why his agent is selling at 0cents. Hope my reservations makes sense

Reply:
Sure, makes sense. In this case I'll leave just the error log, the researcher can then act accordingly indeed. Keep in mind though that the same applies for the UI, that's why we cannot raise an exception here, it would crash the simulation run. An error log looks fine though